### PR TITLE
Realization of extended clickable area.

### PR DESCRIPTION
**Pull request by GreyMS**
Suitable for small lonely objects, which must be clicked.
For activation add "#define USE_LV_EXTENDED_CLICK_AREA 1" to lv_conf.h

### CODE REVIEW

File: src/lv_core/lv_obj.c [hunk 1]
```diff
@@ -643,6 +683,24 @@ void lv_obj_set_size(lv_obj_t * obj, lv_coord_t w, lv_coord_t h)
...
 }
 
+#if USE_LV_EXTENDED_CLICK_AREA
+/**
+ * Set the size of an extended clickable area
+ * @param obj pointer to an object
+ * @param w extended width to both sides
+ * @param h extended height to both sides
+ */
+void lv_obj_set_ext_paddinds(lv_obj_t * obj, lv_coord_t w, lv_coord_t h)
```
Comment (embeddedt) [inline src/lv_core/lv_obj.c]: I think this is a spelling error.

File: src/lv_core/lv_obj.h [hunk 2]
```diff
@@ -337,6 +341,16 @@ void lv_obj_set_y(lv_obj_t * obj, lv_coord_t y);
...
 void lv_obj_set_size(lv_obj_t * obj, lv_coord_t w, lv_coord_t h);
 
+#if USE_LV_EXTENDED_CLICK_AREA
+/**
+ * Set the size of an extended clickable area
+ * @param obj pointer to an object
+ * @param w extended width to both sides
+ * @param h extended height to both sides
+ */
+void lv_obj_set_ext_paddinds(lv_obj_t * obj, lv_coord_t w, lv_coord_t h);
```
Comment (embeddedt) [inline src/lv_core/lv_obj.h]: Same spelling error as above.

File: src/lv_core/lv_obj.c [hunk 3]
```diff
@@ -55,6 +55,20 @@ static const void * event_act_data;           /*Stores the data passed to the ev
...
  *      MACROS
  **********************/
+#if USE_LV_EXTENDED_CLICK_AREA
+/**
+ * Update coordinates of extended clickable area from object's coordinates and ext_paddings
+ * @param coords coordinates of an object
+ * @param ext_coords extended coordinates, which will be updated
+ * @param paddings paddings of extended clickable area
+ */
+#define UPDATE_EXT_COORDS(coords, ext_coords, paddings) do{\
```
Comment (embeddedt) [inline src/lv_core/lv_obj.c]: This should be made into a `static` or `static inline` function.

**Review by embeddedt — CHANGES_REQUESTED** [review]

File: src/lv_core/lv_obj.c (hunk 1 above)
Comment (GreyMS) [reply src/lv_core/lv_obj.c]: Yes, of course. Didn't noticed.

File: src/lv_core/lv_obj.c (hunk 3 above)
Comment (GreyMS) [reply src/lv_core/lv_obj.c]: ok, i'll change to inline

**Review by embeddedt — APPROVED** [review]

**Comment by kisvegabor** [comment]
Thank you for the PR!

I have only one concern: it needs 16 extra bytes for each object: 
```c
#if USE_LV_EXTENDED_CLICK_AREA
    lv_area_t ext_coords;
    lv_area_t ext_paddings;
#endif
```

**ext_coords**
It would be possible to calculate it on the fly into a local variable in `indev_search_obj`. This way `update_ext_coords` won't be required.

**Ext paddings**
We need to discuss if it's really required to have all the 4 coordinates.  My current view is that even a single `uint8_t` or `lv_coord_t` variable would be enough (which is applied in all 4 directions) in most of the cases. What do you think?

**Comment by GreyMS** [comment]
I thought about extra memory, but in case of calculation on the fly it will add additional four arithmetic operations at each point comparison for each object. I don’t know what is better: save memory or save cpu. I can add additional option, something like ‘USE_LV_EXTENDED_AREA_TINY’ and realize alternative mechanism with calculation on the fly.
About paddings: I think that horizontal and vertical paddings still needed, because we can have wide but low or narrow but tall widgets, which will need this feature. But paddings can really be stored in unit8_t, so both values will occupy only 16 extra bits.
I will update this PR today.

**Comment by GreyMS** [comment]
I made changes according to the ideas mentioned above. Also I updated my branch to be sure there are no merge conflicts with other commits in origin dev-6.0.

**Comment by embeddedt** [comment]
Theoretically, I think we should allow padding values beyond `255`. Not everyone is on a small screen; there could be some case where someone would want to have a button with a huge extended click area.

**Comment by GreyMS** [comment]
I can change paddings to lv_point instead of two uint8_t. In this case they will occupy 32 bits instead of 16, but will allow any extended click area size.
On the other hand, this feature will likely be used by small amount of objects, but size of all objects will be increased. And 510 pixels rectangle above real object size is rather big even for HD screens.
So, should I change paddings to lv_coord_t ?

**Comment by kisvegabor** [comment]
`USE_LV_EXTENDED_AREA_TINY` is a good idea. But in this case, we can make it really tiny any use only one `lv_coord_t` for all directions.

I prefer calculating the new area on the fly not only because of memory reasons but because it's less error-prone. For example, the new area needs to be calculated in `lv_obj_set_pos` and `lv_obj_set_size` too. However, e.g. `lv_cont` with `fit` [sets the size directly](https://github.com/littlevgl/lvgl/blob/dev-6.0/src/lv_objx/lv_cont.c#L693) not through `lv_obj_set_size`. So there are some places where the coordinates can change and it's easy to miss one.

`lv_conf_checker` is autogenerated from `lv_conf_templ.h` using `scrips/lv_conf_checker.py` so you don't need to modify it directly.

I've run `clang-format` on the source code for the first time which introduced some conflicts. Fortunately, these seem easy to fix.

**Comment by GreyMS** [comment]
I want to PR one more feature. How should I organize it? Create separate branch and try it PR it? Or wait the approvement of current feature and create one more PR after that? Or continue to use this branch and just add info about improvements in this thread?

**Comment by GreyMS** [comment]
Also I didn't know about the places, where coordinates can be modified directly. Indeed it will ruin mechanics of precalculated extended area in these cases. I doesn't delete this part entirely but really it will be better to use tiny version by default. 

I still think that two separate uint8_t paddings is better than one uint16_t for not square objects and use the same amount of memory.

**Comment by embeddedt** [comment]
> I want to PR one more feature. How should I organize it? Create separate branch and try it PR it? Or wait the approvement of current feature and create one more PR after that? Or continue to use this branch and just add info about improvements in this thread?

If it's related to this PR, just add it in. Otherwise create a separate branch.

**Comment by kisvegabor** [comment]
For TINY I agree 2 `uint8_t` is fine.

I've added run time check and renamed things a little bit for consistency.

**Comment by kisvegabor** [comment]
I solved the new conflicts.
@embeddedt Do you also agree with the current version of this feature?

**Review by embeddedt — APPROVED** [review]

**Comment by embeddedt** [comment]
@kisvegabor Looks fine to me!

**Comment by kisvegabor** [comment]
Okay, I merge it.

@GreyMS Thank you for this update! Please, create a new PR if you have new improvement ideas.

**Pull request merged**